tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc

```diff
@@ -0,0 +1,571 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.h"
+
+#include "google/cloud/storage/client.h"
+
+#include <string>
+#include <memory>
+#include <fstream>
+#include <iostream>
+
+#include <stdio.h>
+
+#include "tensorflow/c/tf_status.h"
+
+// Implementation of a filesystem for GCS environments.
+// This filesystem will support `gs://` URI scheme.
+
+// We can cast `google::cloud::StatusCode` to `TF_Code` because they have the same integer values.
+// See https://github.com/googleapis/google-cloud-cpp/blob/6c09cbfa0160bc046e5509b4dd2ab4b872648b4a/google/cloud/status.h#L32-L52
+#define TF_SetStatusFromGCSStatus(gcs_status, status) TF_SetStatus(status, static_cast<TF_Code>(gcs_status.code()), gcs_status.message().c_str())
+#define TF_ReturnIfError(status) if(TF_GetCode(status) != TF_OK) return
+
+static void* plugin_memory_allocate(size_t size) { return calloc(1, size); }
+static void plugin_memory_free(void* ptr) { free(ptr); }
+
+void ParseGCSPath(const char* fname, bool object_empty_ok, char** bucket, char** object, TF_Status* status) {
+  size_t scheme_index = strcspn(fname, "://");
+  char* scheme = (char*) malloc(scheme_index + 1);
+  sprintf(scheme, "%.*s", (int)scheme_index, fname);
+  scheme[scheme_index] = '\0';
+  if(strcmp(scheme, "gs")) {
+    TF_SetStatus(status, TF_INVALID_ARGUMENT, "GCS path doesn't start with 'gs://'.");
+    return;
+  }
+
+  size_t bucket_index = strcspn(fname + scheme_index + 3, "/");
+  if(!bucket_index) {
+    TF_SetStatus(status, TF_INVALID_ARGUMENT, "GCS path doesn't contain a bucket name.");
+    return;
+  }
+  *bucket = (char*) malloc(bucket_index + 1);
+  sprintf(*bucket, "%.*s", (int) bucket_index, fname + scheme_index + 3);
+  (*bucket)[bucket_index] = '\0';
+
+  size_t object_index = strlen(fname + scheme_index + 3 + bucket_index + 1);
+  if(object_index == 0) {
+    if(object_empty_ok) {
+      TF_SetStatus(status, TF_OK, "");
+      *object = nullptr;
+      return;
+    }
+    else {
+      TF_SetStatus(status, TF_INVALID_ARGUMENT, "GCS path doesn't contain an object name.");
+      return;
+    }
+  }
+  *object = (char*) malloc(object_index + 1);
+  sprintf(*object, "%.*s", (int) object_index, fname + scheme_index + 3 + bucket_index + 1);
+  (*object)[object_index] = '\0';
+
+  free(scheme);
+  TF_SetStatus(status, TF_OK, "");
+}
+
+std::shared_ptr<std::fstream> CreateTempFile(char** temp_path_) {
+	*temp_path_ = (char*) malloc(L_tmpnam);
+	*temp_path_ = tmpnam(*temp_path_);
```

Comment:
This is unsafe: http://man7.org/linux/man-pages/man3/tmpnam.3.html#BUGS

Reply:
I haven't noticed that. `tmpfile()` seems to be a good alternative